### PR TITLE
Various enhancements for using server-side component 

### PR DESCRIPTION
**Pull request by ThePrez**
- Add ability to explicitly ask server component for version and build date
- Add support for server-side paging (database cursor-based) for large result sets
- Add prepared statement and parameter marker support
- Allow CL commands to be run in the server database job
- Allow for enabling of server component trace at various levels (data stored in file or in-memory)
- Allow for enabling of JTOpen trace at various levels (data stored in file or in-memory)
- Allow for retrieval of trace data
- Simple tests for the above

### CODE REVIEW

File: src/connection/query.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    this.isPrepared = (undefined !== parms);
+    this.parameters = parms;
+    this.sql = query;
+    Query.globalQueryList.push(this);
+  }
+  public static byId(id: string) {
+    if (undefined === id || '' === id) {
+      return undefined;
+    }
+    for (let srch of Query.globalQueryList) {
```
Comment (worksofliam) [inline src/connection/query.ts:25]: Perhaps:

```
const src = Query.globalQueryList.find(query => query.correlationId === id)
```

File: src/connection/query.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      if (srch.correlationId === id) {
+        return srch;
+      }
+    } return undefined;
+  }
+  public async run(rowsToFetch: number = this.rowsToFetch): Promise<QueryResult<T>> {
+    if (this.hasBeenRunAlready) {
```
Comment (worksofliam) [inline src/connection/query.ts:32]: I think calling `run` twice needs a test case. I personally think this should return an error.

**Comment by worksofliam** [comment]
@ThePrez Any chance you can update the description of this PR with a high level list of changes? Cheers!

**Comment by ThePrez** [comment]
Sure thing. I just edited the issue itself with bullets

File: src/connection/query.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import { CLCommandResult, JobLogEntry, QueryResult, ServerResponse } from "./types";
+export enum QueryState {
+  NOT_YET_RUN = 1,
+  RUN_MORE_DATA_AVAILABLE = 2,
+  RUN_DONE = 3,
+  ERROR = 4
+}
+export class Query<T> {
+
+  private static id_ctr: number = 43;
```
Comment (worksofliam) [inline src/connection/query.ts:11]: This could have a better name.

File: src/connection/sqlJob.ts [hunk 4]
```diff
@@ -102,91 +103,61 @@ export class SQLJob {
     return connectResult;
   }
 
-  async query<T>(sql: string, parameters?: (number | string)[]): Promise<T[]> {
-    const hasParms = (parameters && parameters.length > 0);
+  query<T>(isCL, sql: string, parms: any[] = undefined): Query<T> {
```
Comment (worksofliam) [inline src/connection/sqlJob.ts:106]: No type on `isCL`

Comment (worksofliam) [reply src/connection/sqlJob.ts:106]: I think it should look like this:

`query<T>(query: string, options: QueryOptons): Query<T>`.

`QueryOptions` should have `parms` and `isCL`

File: src/connection/manager.ts [hunk 5]
```diff
@@ -86,11 +87,11 @@ export class SQLJobManager {
     return (this.selectedJob >= 0);
   }
 
-  runSQL<T>(query: string): Promise<T[]> {
+  async runSQL<T>(query: string): Promise<T[]> {
     const selected = this.jobs[this.selectedJob]
     if (SQLJobManager.jobSupport && selected) {
-      return selected.job.query(query) as Promise<T[]>;
-
+      let results = await selected.job.query<T>(false, query).run();
```
Comment (worksofliam) [inline src/connection/manager.ts:93]: We need to call `run` with a high value here.

File: src/connection/query.ts (hunk 3 above)
Comment (ThePrez) [reply src/connection/query.ts]: resolved at https://github.com/halcyon-tech/vscode-db2i/pull/23/commits/ba523bac2febfba9fdaa67d613a77e721b212df3

File: src/connection/manager.ts (hunk 5 above)
Comment (ThePrez) [reply src/connection/manager.ts]: Resolved at https://github.com/halcyon-tech/vscode-db2i/pull/23/commits/c8fc13db33e849e64ca40f67982004830148b867

File: src/connection/sqlJob.ts (hunk 4 above)
Comment (ThePrez) [reply src/connection/sqlJob.ts]: resolved at https://github.com/halcyon-tech/vscode-db2i/pull/23/commits/59be615be0a6a012724922f76b02eec049112866

File: src/connection/manager.ts [hunk 6]
```diff
@@ -86,11 +87,14 @@ export class SQLJobManager {
...
 
-  runSQL<T>(query: string): Promise<T[]> {
+  async runSQL<T>(query: string): Promise<T[]> {
     const selected = this.jobs[this.selectedJob]
     if (SQLJobManager.jobSupport && selected) {
-      return selected.job.query(query) as Promise<T[]>;
-
+      // 2147483647 is NOT arbitrary. On the server side, this is processed as a Java
+      // int. This is the largest number available without overflow
+      let rowsToFetch = 2147483647;
```
Comment (worksofliam) [inline src/connection/manager.ts:95]: Can this be a const?

Comment (ThePrez) [reply src/connection/manager.ts]: done, https://github.com/halcyon-tech/vscode-db2i/pull/23/commits/b81039529a63d4a70124bc8e7a5d945e8eda863c

**Pull request merged**